### PR TITLE
Substitute types before checking inlining compatibility.

### PR DESCRIPTION
**Pull request by cjgillot**
Addresses https://github.com/rust-lang/rust/issues/112332 and https://github.com/rust-lang/rust/issues/113781

I don't have a minimal test, but I this seems to remove the ICE locally.

This whole pre-inlining validation mirrors the "real" MIR validation pass to verify that inlined MIR will still pass validation.
The debuginfo loop is added because MIR validation check projections in debuginfo.
Likewise, MIR validation only checks `is_subtype`, so there is no reason for a stronger check.

The types were not being substituted in `check_equal`, so we were not bailing out of inlining if the substituted MIR callee body would not pass validation.

### CODE REVIEW

File: compiler/rustc_mir_transform/src/inline.rs [hunk 1]
```diff
@@ -440,6 +440,10 @@ impl<'tcx> Inliner<'tcx> {
             validation: Ok(()),
         };
 
+        for var_debug_info in callee_body.var_debug_info.iter() {
```
Comment (compiler-errors) [inline compiler/rustc_mir_transform/src/inline.rs:443]: What's this for?

File: compiler/rustc_mir_transform/src/inline.rs [hunk 2]
```diff
@@ -847,7 +851,16 @@ impl<'tcx> Visitor<'tcx> for CostChecker<'_, 'tcx> {
...
+                // Fast path if there is nothing to substitute.
+                if ty == f_ty {
+                    return;
+                }
+                let ty = this.instance.subst_mir(this.tcx, ty::EarlyBinder::bind(&ty));
+                let f_ty = this.instance.subst_mir(this.tcx, ty::EarlyBinder::bind(&f_ty));
+                if ty == f_ty {
+                    return;
+                }
+                if !util::is_subtype(this.tcx, this.param_env, ty, f_ty) {
```
Comment (compiler-errors) [inline compiler/rustc_mir_transform/src/inline.rs:863]: Makes sense, but why are we changing this from `is_equal_up_to_subtyping`? I guess do we only want to check that `ty <: f_ty`?

**Comment by cjgillot** [comment]
I should have put more context in the PR comment.
This whole pre-inlining validation mirrors the "real" MIR validation pass to verify that inlined MIR will still pass validation.
The debuginfo loop is added because MIR validation check projections in debuginfo.
Likewise, MIR validation only checks `is_subtype`, so there is no reason for a stronger check.

This ad-hoc validation should reuse the existing MIR validation code, and not brew its own, but that's for another PR.

**Review by compiler-errors — APPROVED** [review]

**Comment by cjgillot** [comment]
I'm nominating for backport, as both ICEs affect stable 1.71.0.

File: compiler/rustc_mir_transform/src/inline.rs (hunk 1 above)
Comment (pnkfelix) [reply compiler/rustc_mir_transform/src/inline.rs:443]: (answered in: https://github.com/rust-lang/rust/pull/113802#issuecomment-1640709427 )

**Comment by apiraino** [comment]
Beta + stable backports approved as per compiler team [on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202023-07-27/near/379178042)

@rustbot label +beta-accepted +stable-accepted

File: compiler/rustc_mir_transform/src/inline.rs (hunk 2 above)
Comment (RalfJung) [inline compiler/rustc_mir_transform/src/inline.rs:863]: Would be good to add a comment pointing to the issue(s), to give some context to this check.

**Pull request merged**